src/os/bluestore/HybridAllocator_impl.h

```diff
@@ -0,0 +1,186 @@
+    int64_t res2 = 0;
+    // try alternate allocator
+    if (!primary_first) {
+      res2 = T::_allocate(want - res, unit, max_alloc_size, hint, extents);
+    } else if (bmap_alloc) {
+      res2 =
+        bmap_alloc->allocate(want - res, unit, max_alloc_size, hint, extents);
+    }
+    if (res2 >= 0) {
+      res += res2;
+    } else {
+      // allocator shouldn't return new extents on error
+      ceph_assert(orig_size == extents->size());
+    }
+  }
+  return res ? res : -ENOSPC;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::dump()
+{
+  std::lock_guard l(T::get_lock());
+  T::_dump();
+  if (bmap_alloc) {
+    bmap_alloc->dump();
+  }
+  dout(0) << __func__
+    << " avl_free: " << T::_get_free()
+    << " bmap_free: " << (bmap_alloc ? bmap_alloc->get_free() : 0)
+    << dendl;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::init_rm_free(uint64_t offset, uint64_t length)
+{
+  if (!length)
+    return;
+  std::lock_guard l(T::get_lock());
+  dout(10) << __func__ << std::hex
+    << " offset 0x" << offset
+    << " length 0x" << length
+    << std::dec << dendl;
+  T::_try_remove_from_tree(offset, length,
+    [&](uint64_t o, uint64_t l, bool found) {
+      if (!found) {
+        if (bmap_alloc) {
+          bmap_alloc->init_rm_free(o, l);
+        } else {
+          derr << __func__ << " lambda " << std::hex
+            << "Uexpected extent: "
+            << " 0x" << o << "~" << l
+            << std::dec << dendl;
+          ceph_assert(false);
+        }
+      }
+    });
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::_spillover_range(uint64_t start, uint64_t end)
+{
+  auto size = end - start;
+  dout(20) << __func__
+    << std::hex << " "
+    << start << "~" << size
+    << std::dec
+    << dendl;
+  ceph_assert(size);
+  if (!bmap_alloc) {
+    dout(1) << __func__
+      << " constructing fallback allocator"
+      << dendl;
+    bmap_alloc = new BitmapAllocator(T::get_context(),
+      T::get_capacity(),
+      T::get_block_size(),
+      T::get_name() + ".fallback");
+  }
+  bmap_alloc->init_add_free(start, size);
+}
+
+template <typename T>
+int64_t HybridAllocatorBase<T>::_spillover_allocate(uint64_t want,
+  uint64_t unit,
+  uint64_t max_alloc_size,
+  int64_t  hint,
+  PExtentVector* extents)
+{
+  return __allocate_or_rollback(false,
+    want,
+    unit,
+    max_alloc_size,
+    hint,
+    extents);
+}
+
+template <typename T>
+int64_t HybridAllocatorBase<T>::__allocate_or_rollback(
+  bool primary,
+  uint64_t want,
+  uint64_t unit,
+  uint64_t max_alloc_size,
+  int64_t  hint,
+  PExtentVector* extents)
+{
+  int64_t res = 0;
+  ceph_assert(extents);
+  // preserve original 'extents' vector state
+  auto orig_size = extents->size();
+  if (primary) {
+    res = T::_allocate(want, unit, max_alloc_size, hint, extents);
+  } else if (bmap_alloc) {
+    res = bmap_alloc->allocate(want, unit, max_alloc_size, hint, extents);
+  }
+  if (res < 0) {
```

Comment:
In the case that `res < 0` it would be r==-ENOSPC.
I think you had in mind `res < want`.

Reply:
No. Here res < 0 means an error in allocator while res < want (and >= 0) is a valid state - we just don't have enough space available (bit have allocated some). So for the former case we want to release everything allocated and return 0 and for the latter - just return what we've got.